### PR TITLE
ReferenceLine reads inputs from context instead of props

### PR DESCRIPTION
**Pull request by PavelVanecek**
Which also means it no longer needs to be cloned!

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

No more cloning

## How Has This Been Tested?

Storybooks, npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] I have added a storybook story or extended an existing story to show my changes
- [X] All new and existing tests passed.

### CODE REVIEW

File: src/cartesian/ReferenceLine.tsx [hunk 1]
```diff
@@ -150,7 +138,19 @@ export const getEndPoints = (
...
+  const [{ xAxisMap, yAxisMap, viewBox, clipPathId }] = useContext(ChartLayoutContext);
+  if (!clipPathId || !xAxisMap || !yAxisMap || !viewBox) {
+    return null;
+  }
+  const xAxis = xAxisMap[xAxisId];
+  const yAxis = yAxisMap[yAxisId];
+
+  warn(
+    xAxis != null,
+    `Could not find xAxis by id ${xAxisId} [${typeof xAxisId}]. Available xAxis are: ${Object.keys(xAxisMap)}`,
```
Comment (ckifer) [inline src/cartesian/ReferenceLine.tsx:152]: is this an invariant or a warning? If we don't have an xAxis isnt that an error?

**Review by ckifer — APPROVED** [review]
Absolutely huge for the first step here! Going to pull down and play with it a bit

File: src/cartesian/ReferenceLine.tsx [hunk 2]
```diff
@@ -150,7 +138,19 @@ export const getEndPoints = (
 };
 
 export function ReferenceLine(props: Props) {
-  const { x: fixedX, y: fixedY, segment, xAxis, yAxis, shape, className, alwaysShow, clipPathId } = props;
+  const { x: fixedX, y: fixedY, segment, xAxisId, yAxisId, shape, className, alwaysShow } = props;
+
+  const [{ xAxisMap, yAxisMap, viewBox, clipPathId }] = useContext(ChartLayoutContext);
```
Comment (HHongSeungWoo) [inline src/cartesian/ReferenceLine.tsx:143]: What do you think about creating and using the "useChartLayoutContext"???

File: src/chart/generateCategoricalChart.tsx [hunk 3]
```diff
@@ -2304,22 +2338,25 @@ export const generateCategoricalChart = ({
...
+            }}
+            role="region"
+          >
+            <Surface {...attrs} width={width} height={height} title={title} desc={desc} style={FULL_WIDTH_AND_HEIGHT}>
+              {this.renderClipPath()}
+              {renderByOrder(children, this.renderMap)}
+            </Surface>
+            {this.renderLegend()}
+            {this.renderTooltip()}
+            {this.updateLayoutContext()}
```
Comment (HHongSeungWoo) [inline src/chart/generateCategoricalChart.tsx:2357]: I'm concerned about updating the context within the context. 
As it stands, It feels like adding unnecessary rendering behaviors.

Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx:2357]: That's right. What can I do instead to update the context without the double render?

File: src/cartesian/ReferenceLine.tsx (hunk 2 above)
Comment (PavelVanecek) [reply src/cartesian/ReferenceLine.tsx:143]: I don't follow. Could you please rephrase your question?

File: src/cartesian/ReferenceLine.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/cartesian/ReferenceLine.tsx:152]: The component will throw few lines below anyway, might just as well make it invariant.

File: src/chart/generateCategoricalChart.tsx (hunk 3 above)
Comment (HHongSeungWoo) [reply src/chart/generateCategoricalChart.tsx:2357]: Is it acceptable to directly pass value to the Provider without using useState?

File: src/cartesian/ReferenceLine.tsx (hunk 2 above)
Comment (HHongSeungWoo) [reply src/cartesian/ReferenceLine.tsx:144]: If implemented as `export useChartLayoutContext = () => useContext(ChartLayoutContext);`
it seems like there could be a few small benefits.

Comment (PavelVanecek) [reply src/cartesian/ReferenceLine.tsx:144]: Sure I can do that. I'm wondering what are the benefits?

File: src/chart/generateCategoricalChart.tsx (hunk 3 above)
Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx:2357]: I'll try that

File: src/cartesian/ReferenceLine.tsx (hunk 2 above)
Comment (HHongSeungWoo) [reply src/cartesian/ReferenceLine.tsx:144]: Hmm... It seems that the import statements will be cleaned up and the implementation will be hidden, so it will be free to modify the internal implementation.

File: src/cartesian/ReferenceLine.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/cartesian/ReferenceLine.tsx]: Done

File: src/cartesian/ReferenceLine.tsx (hunk 2 above)
Comment (PavelVanecek) [reply src/cartesian/ReferenceLine.tsx]: Done

File: src/chart/generateCategoricalChart.tsx (hunk 3 above)
Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx]: Done, what do you think @HHongSeungWoo ?

**Comment by PavelVanecek** [comment]
I will add more tests before merging, if we decide that this is good approach.

File: src/chart/generateCategoricalChart.tsx (hunk 3 above)
Comment (HHongSeungWoo) [reply src/chart/generateCategoricalChart.tsx]: I think it's good!

**Review by HHongSeungWoo — APPROVED** [review]

File: src/context/layoutContext.tsx
Comment (nikolasrieble) [inline src/context/layoutContext.tsx:1]: This file exports ChartLayoutContext but is named layoutContext

Lets use either word consistently: LayoutContext or ChartLayoutContext

File: storybook/stories/API/cartesian/ReferenceLine.stories.tsx [hunk 4]
```diff
@@ -130,6 +130,6 @@ export const API = {
     stroke: 'blue',
     strokeWidth: 2,
     strokeDasharray: '4 1',
-    label: { fill: 'red', fontSize: 20 },
+    label: 'My example label',
```
Comment (nikolasrieble) [inline storybook/stories/API/cartesian/ReferenceLine.stories.tsx:133]: Is this an unrelated fix? Do we need this change?

File: test/cartesian/ReferenceLine/ReferenceLine.spec.tsx [hunk 5]
```diff
@@ -296,4 +296,62 @@ describe('<ReferenceLine />', () => {
...
+          width={1100}
+          height={250}
+          barGap={2}
+          barSize={6}
+          data={data}
+          margin={{ top: 20, right: 60, bottom: 0, left: 20 }}
+        >
+          <XAxis dataKey="name" />
+          <YAxis tickCount={7} />
+          <Bar dataKey="uv" />
```
Comment (nikolasrieble) [inline test/cartesian/ReferenceLine/ReferenceLine.spec.tsx:314]: Do we need the bar for this test? Ideally the test should be as simple as possible.

File: test/cartesian/ReferenceLine/ReferenceLine.spec.tsx [hunk 6]
```diff
@@ -296,4 +296,62 @@ describe('<ReferenceLine />', () => {
...
+
+  test('throws when passed in invalid xAxisId', () => {
+    const viewBox: CartesianViewBox = { x: 1, y: 2 };
+    expect(() => {
+      render(
+        <BarChart
+          width={1100}
+          height={250}
+          barGap={2}
+          barSize={6}
```
Comment (nikolasrieble) [inline test/cartesian/ReferenceLine/ReferenceLine.spec.tsx:308]: Can we further simplify this test by removing unneeded properties? These seem unrelated.

File: test/cartesian/ReferenceLine/ReferenceLine.spec.tsx [hunk 7]
```diff
@@ -296,4 +296,62 @@ describe('<ReferenceLine />', () => {
...
+          margin={{ top: 20, right: 60, bottom: 0, left: 20 }}
+        >
+          <XAxis dataKey="name" />
+          <YAxis tickCount={7} />
+          <Bar dataKey="uv" />
+          <ReferenceLine
+            y={20}
+            stroke="#666"
+            ifOverflow="visible"
+            viewBox={viewBox}
```
Comment (nikolasrieble) [inline test/cartesian/ReferenceLine/ReferenceLine.spec.tsx:319]: I assume these are also not needed

File: src/context/layoutContext.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import { CartesianViewBox } from '../util/types';
+import { XAxisProps, YAxisProps } from '../index';
+
+export type XAxisMap = {
+  [axisId: string]: XAxisProps;
+};
+
+export type YAxisMap = {
+  [axisId: string]: YAxisProps;
+};
```
Comment (nikolasrieble) [inline src/context/layoutContext.tsx:11]: Are these types specific to the layoutContext? I expect them to be used in many other places - should we put them into a more generic location to increase reusability / visibility?

**Review by nikolasrieble — CHANGES_REQUESTED** [review]
Great step, thank you for taking this on. Please simplify tests as much as possible and consider renaming layoutContext

File: storybook/stories/API/cartesian/ReferenceLine.stories.tsx (hunk 4 above)
Comment (PavelVanecek) [reply storybook/stories/API/cartesian/ReferenceLine.stories.tsx:133]: Yep, the label is meant to be a text.

File: src/cartesian/ReferenceLine.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/cartesian/ReferenceLine.tsx]: To confirm I wrote these tests on `master` branch:

```  test('throws an error when rendered without any axes', () => {
    render(
      <BarChart width={1100} height={250}>
        <ReferenceLine y={20} />
      </BarChart>,
    );
  });

  test('throws an error when rendered without YAxis', () => {
    render(
      <BarChart width={1100} height={250}>
        <XAxis />
        <ReferenceLine y={20} />
      </BarChart>,
    );
  });

  test('throws an error when rendered without XAxis', () => {
    render(
      <BarChart width={1100} height={250}>
        <YAxis />
        <ReferenceLine y={20} />
      </BarChart>,
    );
  });
```

And indeed they all do fail:

image

image

So a side effect of this PR is a better error message.

I wonder if there is ask to have the ReferenceLine work with one axis defined. I will browse through open issues later.

File: src/context/layoutContext.tsx (hunk 8 above)
Comment (PavelVanecek) [reply src/context/layoutContext.tsx]: Done

File: test/cartesian/ReferenceLine/ReferenceLine.spec.tsx (hunk 7 above)
Comment (PavelVanecek) [reply test/cartesian/ReferenceLine/ReferenceLine.spec.tsx]: Done

File: test/cartesian/ReferenceLine/ReferenceLine.spec.tsx (hunk 6 above)
Comment (PavelVanecek) [reply test/cartesian/ReferenceLine/ReferenceLine.spec.tsx]: Done

File: test/cartesian/ReferenceLine/ReferenceLine.spec.tsx (hunk 5 above)
Comment (PavelVanecek) [reply test/cartesian/ReferenceLine/ReferenceLine.spec.tsx]: Done

File: src/context/layoutContext.tsx
Comment (PavelVanecek) [reply src/context/layoutContext.tsx:1]: Done

**Comment by PavelVanecek** [comment]
Few more tests I would like to add

- [ ] Test for rendering with context alone, outside of chart
- [ ] Test for rendering outside of context
- [ ] Test for rendering in nested child component

**Review by ckifer — APPROVED** [review]
These tests are incredible

I say lets ship this and see how things go! 

Only thing I will say is that we still technically support React 15... This breaks support for React < 16.3 which is when Context came out. I would think most people still on R 15 aren't upgrading anything anymore but just calling it out

**Comment by ckifer** [comment]
Actually maybe I will do a release and then merge this...

**Comment by PavelVanecek** [comment]
Right, I didn't consider that. And I used a hook and hooks are 16.8+ even.

**Comment by ckifer** [comment]
Oh wow, did that happen and I didn't notice? I know R15 was there for a while... or maybe that was in react-smooth. 

Either way that makes me feel better. 

Going to do a patch release today then I'll merge this

**Comment by ckifer** [comment]
Aight lets do it!
